### PR TITLE
Bugfix `CATKEVerticalDiffusivity` locator functions + other fixes

### LINKED ISSUE #3047: `Callable` tridiagonal coefficients produces catastrophic loss of performance
Currently `BatchedTridiagonalSolver` supports `Callable` tridiagonal coefficients via

https://github.com/CliMA/Oceananigans.jl/blob/bcc34f07b3f949ea6fb34c7814f4b856d24924c2/src/Solvers/batched_tridiagonal_solver.jl#L58-L59

However, as noted on #3030 this can produce a catastrophic loss of performance. Therefore, this support should be discontinued.

cc @simone-silvestri

### PR DESCRIPTION
**Pull request by glwagner**
Plus fixing vertically implicit diffusion for partial cells and a few other things.

Closes https://github.com/CliMA/Oceananigans.jl/issues/3064
Closes #3047

### CODE REVIEW

**Comment by glwagner** [comment]
Using the stencil on `main`:

image

This PR:

image

This problem is symmetric. Before this PR the solution is not symmetric, but after it is.

However, this could also be a bug in CATKE (or something).

**Comment by glwagner** [comment]
Code:

```julia
using Oceananigans
using Oceananigans.Units
using Oceananigans.TurbulenceClosures: CATKEVerticalDiffusivity
using GLMakie
using Printf

Lz = 128
Δz = 8
zb = -(Lz + Δz) / 2
zt = +(Lz + Δz) / 2
z = collect(zb:Δz:zt)
Nz = length(z) - 1

grid = RectilinearGrid(size=Nz, z=z, topology=(Flat, Flat, Bounded))

closure = CATKEVerticalDiffusivity()
model = HydrostaticFreeSurfaceModel(; grid, buoyancy=BuoyancyTracer(), tracers=(:b, :e), closure)

Δ = 16
N² = 1e-6
Ri = 0.1 # N² / S²
dUdz = sqrt(N² / Ri)
@show U = dUdz * Δ
uᵢ(x, y, z) = U * tanh(z / Δ)
bᵢ(x, y, z) = N² * z
set!(model, u=uᵢ, b=bᵢ, e=1e-6)

u = model.velocities.u
for k = 1:Nz
    @info @sprintf("u[%d] = %.2f", k, u[1, 1, k])
end

simulation = Simulation(model, Δt=10minutes, stop_time=24hours)

bt = []
ut = []
et = []
κct = []
κut = []
Rit = []

b = model.tracers.b
u = model.velocities.u
e = model.tracers.e
κc = model.diffusivity_fields.κᶜ
κu = model.diffusivity_fields.κᵘ
Ri = Field(∂z(b) / ∂z(u)^2)

function collect_data(sim)
    compute!(Ri)
    push!(Rit, deepcopy(interior(Ri, 1, 1, :)))
    push!(bt,  deepcopy(interior(b, 1, 1, :)))
    push!(ut,  deepcopy(interior(u, 1, 1, :)))
    push!(et,  deepcopy(interior(e, 1, 1, :)))
    push!(κct, deepcopy(interior(κc, 1, 1, :)))
    push!(κut, deepcopy(interior(κu, 1, 1, :)))

    return nothing
end

t = 0:10minutes:simulation.stop_time
Nt = length(t)

simulation.callbacks[:dc] = Callback(collect_data, SpecifiedTimes(t))

run!(simulation)

fig = Figure(resolution=(800, 600))

axR0 = Axis(fig[1, 1], xlabel="Ri⁻¹", ylabel="z (m)")
axe0 = Axis(fig[2, 1], xlabel="e", ylabel="z (m)", xticks=[1e-6, 2e-5])
axκ0 = Axis(fig[3, 1], xlabel="κ", ylabel="z (m)", xticks=[0.0, 2e-3, 4e-3])

axR = Axis(fig[1, 2], xlabel="Time (hr)", ylabel="z (m)", xaxisposition=:top)
axe = Axis(fig[2, 2], ylabel="z (m)")
axκ = Axis(fig[3, 2], xlabel="Time (hr)", ylabel="z (m)")

hidexdecorations!(axe)

hidespines!(axR0, :r, :t)
hidespines!(axe0, :r, :t)
hidespines!(axκ0, :r, :t)

Prt = map((κu, κc) -> κu ./ κc, κut, κct)

bzt = hcat(bt...)'
Rzt = hcat(Rit...)'
ezt = hcat(et...)'
κzt = hcat(κct...)'
Pzt = hcat(Prt...)'

zc = znodes(grid, Center())
zf = znodes(grid, Face())

for n = (7, 49, 24*4+1)
    @show tn = t[n] / hour
    label = @sprintf("t = %d hr", tn)
    lines!(axR0, 1 ./ Rit[n], zf; label)
    ln = lines!(axe0, et[n], zc)
    lines!(axκ0, κct[n], zf, color = ln.color.val, label="κᶜ")
    lines!(axκ0, κut[n], zf, linestyle=:dash, color=ln.color.val, label="κᵘ")
end

Legend(fig[2, 0], axR0)
Legend(fig[3, 0], axκ0, merge=true)

vlines!(axR0, 4, color=(:gray, 0.5), linewidth=4)

cr = contourf!(axR, t ./ hour, zf, 1 ./ Rzt, levels = 0:10, colormap=:viridis)#, extendhigh=true, extendlow=true)
Colorbar(fig[1, 3], cr, label="Ri⁻¹")

cr = contourf!(axe, t ./ hour, zc, ezt, levels=10, colormap=:heat)
Colorbar(fig[2, 3], cr, label="E (m² s⁻²)")

cr = contourf!(axκ, t ./ hour, zf, κzt, levels=10, colormap=:solar)
Colorbar(fig[3, 3], cr, label="κᶜ (m² s⁻¹)")

colsize!(fig.layout, 1, Relative(0.2))

display(fig)
```

File: src/TurbulenceClosures/turbulence_closure_implementations/CATKEVerticalDiffusivities/CATKEVerticalDiffusivities.jl [hunk 1]
```diff
@@ -37,6 +38,8 @@ import Oceananigans.TurbulenceClosures:
     implicit_linear_coefficient,
     viscosity,
     diffusivity,
+    viscosity_location,
+    diffusivity_location,
```
Comment (glwagner) [inline src/TurbulenceClosures/turbulence_closure_implementations/CATKEVerticalDiffusivities/CATKEVerticalDiffusivities.jl:42]: I now suspect this was the problem @navidcy @simone-silvestri

**Comment by simone-silvestri** [comment]
Nice! I ll take a look later

**Comment by glwagner** [comment]
Undoing the change in the vertically implicit diffusion solver, and then fixing hte CATKE bug leads to:

image

This is the diffusion of a shear layer with Ri=0.1 on a stretched vertical grid. The grid is symmetric around z=0 (which helps isolate the fact that the issue was not with the tridiagonal solver, but with CATKE).

**Comment by navidcy** [comment]
How about the tests that fail? Wanna work together to get them done?

File: Manifest.toml
Comment (navidcy) [inline Manifest.toml:1]: add back Manifest?

**Comment by navidcy** [comment]
patch release please

File: Manifest.toml
Comment (glwagner) [reply Manifest.toml:1]: yes

**Comment by glwagner** [comment]
> How about the tests that fail? Wanna work together to get them done?

yes, that's unfortunate, let's!

**Comment by navidcy** [comment]
perhaps test fail because there is no manifest and different KA/CUDA/whatnot versions are being installed.

**Comment by glwagner** [comment]
It's less sensitive to resolution with the fix. Here's some boundary layer solutions at 2, 8, and 16m resolution.

variable_Pr_conv_adj_6_hour_suite_assessment

variable_Pr_conv_adj_72_hour_suite_assessment

**Pull request closed without merging**